diff --git a/common/spl/spl_nand.c b/common/spl/spl_nand.c
index 59f4a84a3..671fff2d4 100644
--- a/common/spl/spl_nand.c
+++ b/common/spl/spl_nand.c
@@ -71,6 +71,7 @@ static int spl_nand_load_element(struct spl_image_info *spl_image,
 	int bl_len = mtd ? mtd->writesize : 1;
 	int err;
 
+	puts("@mah: Loading image header ...\n");
 	err = nand_spl_load_image(offset, sizeof(*header), (void *)header);
 	if (err)
 		return err;
@@ -96,9 +97,11 @@ static int spl_nand_load_element(struct spl_image_info *spl_image,
 		load.read = spl_nand_fit_read;
 		return spl_load_imx_container(spl_image, &load, offset / bl_len);
 	} else {
+		puts("@mah: Parsing image header ...\n");
 		err = spl_parse_image_header(spl_image, header);
 		if (err)
 			return err;
+		printf("@mah: Loading image of size=%d to loadAddr=0x%08lx ...\n", spl_image->size, spl_image->load_addr);
 		return nand_spl_load_image(offset, spl_image->size,
 					   (void *)(ulong)spl_image->load_addr);
 	}
@@ -171,14 +174,17 @@ static int spl_nand_load_image(struct spl_image_info *spl_image,
 #endif
 #endif
 	/* Load u-boot */
+	printf("@mah: Trying to load image from NAND offset=0x%08x ...\n", CONFIG_SYS_NAND_U_BOOT_OFFS);
 	err = spl_nand_load_element(spl_image, spl_nand_get_uboot_raw_page(),
 				    header);
 #ifdef CONFIG_SYS_NAND_U_BOOT_OFFS_REDUND
 #if CONFIG_SYS_NAND_U_BOOT_OFFS != CONFIG_SYS_NAND_U_BOOT_OFFS_REDUND
-	if (err)
+	if (err) {
+		printf("@mah: Trying to load image from NAND redund offset=0x%08x ...\n", CONFIG_SYS_NAND_U_BOOT_OFFS_REDUND);
 		err = spl_nand_load_element(spl_image,
 					    CONFIG_SYS_NAND_U_BOOT_OFFS_REDUND,
 					    header);
+	}
 #endif
 #endif
 	nand_deselect();
